From f88bccc9787f688b50d7df77cbfe8cbf77a1c3c1 Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Date: Tue, 6 Dec 2022 08:59:17 +0200
Subject: [PATCH 01/25] gpio: s32cc: Add get_direction callback

All GPIO hogs listed in device tree are initialized during
gpiochip_add_data_with_key function. Right after, the GPIO
framework will proceed to initialization of all GPIO pins
attached to the controller under initialization based on
the presence of get_direction callback. A missing get_direction
callback will make the framework to assume that all pins have
input direction and will initialize them accordingly. Therefore,
a GPIO hog listed as output pin will be reset to input direction.

Issue: ALB-9588
Upstream-Status: Pending 

Signed-off-by: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/gpio/gpio-siul2-s32cc.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/gpio/gpio-siul2-s32cc.c b/drivers/gpio/gpio-siul2-s32cc.c
index 7e499360d46b..3ae28fb89e38 100644
--- a/drivers/gpio/gpio-siul2-s32cc.c
+++ b/drivers/gpio/gpio-siul2-s32cc.c
@@ -218,7 +218,7 @@ static inline void gpio_set_direction(struct siul2_gpio_dev *dev, int gpio,
 }
 
 static inline enum gpio_dir gpio_get_direction(struct siul2_gpio_dev *dev,
-							int gpio)
+							unsigned int gpio)
 {
 	return test_bit(gpio, dev->pin_dir_bitmap) ? OUT : IN;
 }
@@ -243,6 +243,17 @@ static int siul2_gpio_dir_in(struct gpio_chip *chip, unsigned int gpio)
 	return ret;
 }
 
+static int siul2_gpio_get_dir(struct gpio_chip *chip, unsigned int gpio)
+{
+	struct siul2_gpio_dev *gpio_dev = to_siul2_gpio_dev(chip);
+	enum gpio_dir dir = gpio_get_direction(gpio_dev, gpio);
+
+	if (dir == IN)
+		return GPIO_LINE_DIRECTION_IN;
+
+	return GPIO_LINE_DIRECTION_OUT;
+}
+
 static int siul2_irq_gpio_index(const struct siul2_device_data *platdata,
 				irq_hw_number_t gpio)
 {
@@ -1179,6 +1190,7 @@ static int siul2_gpio_probe(struct platform_device *pdev)
 	gc->free = siul2_gpio_free;
 	gc->direction_output = siul2_gpio_dir_out;
 	gc->direction_input = siul2_gpio_dir_in;
+	gc->get_direction = siul2_gpio_get_dir;
 	gc->owner = THIS_MODULE;
 
 	girq = &gc->irq;
-- 
2.25.1

